docs/docs/reference/gen_notebooks/dspy_prompt_optimization.md

```diff
@@ -0,0 +1,290 @@
+
+:::
+
+
+
+# Optimizing LLM Workflows Using DSPy and Weave
+
+The [BIG-bench (Beyond the Imitation Game Benchmark)](https://github.com/google/BIG-bench) is a collaborative benchmark intended to probe large language models and extrapolate their future capabilities consisting of more than 200 tasks. The [BIG-Bench Hard (BBH)](https://github.com/suzgunmirac/BIG-Bench-Hard) is a suite of 23 most challenging BIG-Bench tasks that can be quite difficult to be solved using the current generation of language models.
+
+This tutorial demonstrates how we can improve the performance of our LLM workflow implemented  on the **causal judgement task** from the BIG-bench Hard benchmark and evaluate our prompting strategies. We will use [DSPy](https://dspy-docs.vercel.app/) for implementing our LLM workflow and optimizing our prompting strategy. We will also use [Weave](../docs/introduction.md) to track our LLM workflow and evaluate our prompting strategies.
+
+## Installing the Dependencies
+
+We need the following libraries for this tutorial:
+
+- [DSPy](https://dspy-docs.vercel.app/) for building the LLM workflow and optimizing it.
+- [Weave](../introduction.md) to track our LLM workflow and evaluate our prompting strategies.
+- [datasets](https://huggingface.co/docs/datasets/index) to access the Big-Bench Hard dataset from HuggingFace Hub.
+
+
+```python
+!pip install -qU dspy-ai weave datasets
+```
+
+Since we'll be using [OpenAI API](https://openai.com/index/openai-api/) as the LLM Vendor, we will also need an OpenAI API key. You can [sign up](https://platform.openai.com/signup) on the OpenAI platform to get your own API key.
+
+
+```python
+import os
+from getpass import getpass
+
+api_key = getpass("Enter you OpenAI API key: ")
+os.environ["OPENAI_API_KEY"] = api_key
+```
+
+## Enable Tracking using Weave
+
+Weave is currently integrated with DSPy, and including [`weave.init`](../docs/reference/python-sdk/weave/index.md) at the start of our code lets us automatically trace our DSPy functions which can be explored in the Weave UI. Check out the [Weave integration docs for DSPy](../docs/guides/integrations/dspy.md) to learn more.
+
+
+```python
+import weave
+
+weave.init(project_name="dspy-bigbench-hard")
+```
+
+In this tutorial, we use a metadata class inherited from [`weave.Model`](../docs/guides/core-types/models.md) to manage our metadata.
+
+
+```python
+class Metadata(weave.Model):
+    dataset_address: str = "maveriq/bigbenchhard"
+    big_bench_hard_task: str = "causal_judgement"
+    num_train_examples: int = 50
+    openai_model: str = "gpt-3.5-turbo"
+    openai_max_tokens: int = 2048
+    max_bootstrapped_demos: int = 8
+    max_labeled_demos: int = 8
+
+
+metadata = Metadata()
```

Comment:
Should we support this pattern for weave.Model?

Reply:
I don't think so, actually. Either have a `predict` method or not. Could be a `weave.Object`, or if there's room to have a `predict` method, that'd be better

Reply:
I'll revert back the changes to the DSPy cookbook in this PR.
Let's continue discussing this in a separate PR.

Reply:
I've addressed this issue in this PR
https://github.com/wandb/weave/pull/2183